File: output/python/tests/04_native_code.py

import math

def helper_function(x):
    """Native helper function defined before system"""
    return x * 2


from typing import Any, Optional, List, Dict, Callable

class NativeCodeFrameEvent:
    def __init__(self, message: str, parameters):
        self._message = message
        self._parameters = parameters


class NativeCodeFrameContext:
    def __init__(self, event: NativeCodeFrameEvent, default_return):
        self.event = event
        self._return = default_return
        self._data = {}


class NativeCodeCompartment:
    def __init__(self, state: str, parent_compartment = None):
        self.state = state
        self.state_args = {}
        self.state_vars = {}
        self.enter_args = {}
        self.exit_args = {}
        self.forward_event = None
        self.parent_compartment = parent_compartment

    def copy(self) -> 'NativeCodeCompartment':
        c = NativeCodeCompartment(self.state, self.parent_compartment)
        c.state_args = self.state_args.copy()
        c.state_vars = self.state_vars.copy()
        c.enter_args = self.enter_args.copy()
        c.exit_args = self.exit_args.copy()
        c.forward_event = self.forward_event
        return c


class NativeCode:
    def __init__(self):
        self._state_stack = []
        self._context_stack = []
        self.__compartment = NativeCodeCompartment("Active")
        self.__next_compartment = None
        __frame_event = NativeCodeFrameEvent("$>", None)
        self.__kernel(__frame_event)

    def __kernel(self, __e):
        # Route event to current state
        self.__router(__e)
        # Process any pending transition
        while self.__next_compartment is not None:
            next_compartment = self.__next_compartment
            self.__next_compartment = None
            # Exit current state
            exit_event = NativeCodeFrameEvent("<$", self.__compartment.exit_args)
            self.__router(exit_event)
            # Switch to new compartment
            self.__compartment = next_compartment
            # Enter new state (or forward event)
            if next_compartment.forward_event is None:
                enter_event = NativeCodeFrameEvent("$>", self.__compartment.enter_args)
                self.__router(enter_event)
            else:
                # Forward event to new state
                forward_event = next_compartment.forward_event
                next_compartment.forward_event = None
                if forward_event._message == "$>":
                    # Forwarding enter event - just send it
                    self.__router(forward_event)
                else:
                    # Forwarding other event - send $> first, then forward
                    enter_event = NativeCodeFrameEvent("$>", self.__compartment.enter_args)
                    self.__router(enter_event)
                    self.__router(forward_event)

    def __router(self, __e):
        state_name = self.__compartment.state
        handler_name = f"_state_{state_name}"
        handler = getattr(self, handler_name, None)
        if handler:
            handler(__e)

    def __transition(self, next_compartment):
        self.__next_compartment = next_compartment

    def compute(self, value: int) -> int:
        __e = NativeCodeFrameEvent("compute", {"value": value})
        __ctx = NativeCodeFrameContext(__e, None)
        self._context_stack.append(__ctx)
        self.__kernel(__e)
        return self._context_stack.pop()._return

    def use_math(self) -> float:
        __e = NativeCodeFrameEvent("use_math", None)
        __ctx = NativeCodeFrameContext(__e, None)
        self._context_stack.append(__ctx)
        self.__kernel(__e)
        return self._context_stack.pop()._return

    def _state_Active(self, __e):
        if __e._message == "compute":
            value = __e._parameters["value"]
            # Native code with local variables
            temp = value + 10
            result = helper_function(temp)
            print(f"Computed: {value} -> {result}")
            self._context_stack[-1]._return = result
            return
        elif __e._message == "use_math":
            # Using imported math module
            result = math.sqrt(16) + math.pi
            print(f"Math result: {result}")
            self._context_stack[-1]._return = result
            return


def main():
    print("=== Test 04: Native Code Preservation ===")
    s = NativeCode()

    # Test native code in handler with helper function
    result = s.compute(5)
    expected = (5 + 10) * 2  # 30
    assert result == expected, f"Expected {expected}, got {result}"
    print(f"compute(5) = {result}")

    # Test imported module usage
    math_result = s.use_math()
    expected_math = math.sqrt(16) + math.pi
    assert abs(math_result - expected_math) < 0.001, f"Expected ~{expected_math}, got {math_result}"
    print(f"use_math() = {math_result}")

    print("PASS: Native code preservation works correctly")

if __name__ == '__main__':
    main()

File: output/javascript/tests/04_native_code.js

function helper_function(x) {
    // Native helper function defined before system
    return x * 2;
}

export class NativeCodeFrameEvent {
    _message;
    _parameters;

    constructor(message, parameters) {
        this._message = message;
        this._parameters = parameters;
    }
}


export class NativeCodeFrameContext {
    event;
    _return;
    _data;

    constructor(event, default_return) {
        this.event = event;
        this._return = default_return;
        this._data = {  };
    }
}


export class NativeCodeCompartment {
    state;
    state_args;
    state_vars;
    enter_args;
    exit_args;
    forward_event;
    parent_compartment;

    constructor(state, parent_compartment = null) {
        this.state = state;
        this.state_args = {  };
        this.state_vars = {  };
        this.enter_args = {  };
        this.exit_args = {  };
        this.forward_event = null;
        this.parent_compartment = parent_compartment;
    }

    copy() {
        const c = new NativeCodeCompartment(this.state, this.parent_compartment);
        c.state_args = {...this.state_args};
        c.state_vars = {...this.state_vars};
        c.enter_args = {...this.enter_args};
        c.exit_args = {...this.exit_args};
        c.forward_event = this.forward_event;
        return c;
    }
}


export class NativeCode {
    _state_stack;
    __compartment;
    __next_compartment;
    _context_stack;

    constructor() {
        this._state_stack = [];
        this._context_stack = [];
        this.__compartment = new NativeCodeCompartment("Active");
        this.__next_compartment = null;
        const __frame_event = new NativeCodeFrameEvent("$>", null);
        this.__kernel(__frame_event);
    }

    __kernel(__e) {
        // Route event to current state
        this.__router(__e);
        // Process any pending transition
        while (this.__next_compartment !== null) {
            const next_compartment = this.__next_compartment;
            this.__next_compartment = null;
            // Exit current state
            const exit_event = new NativeCodeFrameEvent("<$", this.__compartment.exit_args);
            this.__router(exit_event);
            // Switch to new compartment
            this.__compartment = next_compartment;
            // Enter new state (or forward event)
            if (next_compartment.forward_event === null) {
                const enter_event = new NativeCodeFrameEvent("$>", this.__compartment.enter_args);
                this.__router(enter_event);
            } else {
                // Forward event to new state
                const forward_event = next_compartment.forward_event;
                next_compartment.forward_event = null;
                if (forward_event._message === "$>") {
                    // Forwarding enter event - just send it
                    this.__router(forward_event);
                } else {
                    // Forwarding other event - send $> first, then forward
                    const enter_event = new NativeCodeFrameEvent("$>", this.__compartment.enter_args);
                    this.__router(enter_event);
                    this.__router(forward_event);
                }
            }
        }
    }

    __router(__e) {
        const state_name = this.__compartment.state;
        const handler_name = `_state_${state_name}`;
        const handler = this[handler_name];
        if (handler) {
            handler.call(this, __e);
        }
    }

    __transition(next_compartment) {
        this.__next_compartment = next_compartment;
    }

    compute(value) {
        const __e = new NativeCodeFrameEvent("compute", {"value": value});
        const __ctx = new NativeCodeFrameContext(__e, null);
        this._context_stack.push(__ctx);
        this.__kernel(__e);
        return this._context_stack.pop()._return;
    }

    use_math() {
        const __e = new NativeCodeFrameEvent("use_math", null);
        const __ctx = new NativeCodeFrameContext(__e, null);
        this._context_stack.push(__ctx);
        this.__kernel(__e);
        return this._context_stack.pop()._return;
    }

    _state_Active(__e) {
        if (__e._message === "compute") {
            const value = __e._parameters?.["value"];
            // Native code with local variables
            const temp = value + 10;
            const result = helper_function(temp);
            console.log(`Computed: ${value} -> ${result}`);
            this._context_stack[this._context_stack.length - 1]._return = result;
            return;;
        } else if (__e._message === "use_math") {
            // Using Math module
            const result = Math.sqrt(16) + Math.PI;
            console.log(`Math result: ${result}`);
            this._context_stack[this._context_stack.length - 1]._return = result;
            return;;
        }
    }
}

function main() {
    console.log("=== Test 04: Native Code Preservation ===");
    const s = new NativeCode();

    // Test native code in handler with helper function
    const result = s.compute(5);
    const expected = (5 + 10) * 2;  // 30
    if (result !== expected) {
        throw new Error(`Expected ${expected}, got ${result}`);
    }
    console.log(`compute(5) = ${result}`);

    // Test Math module usage
    const mathResult = s.use_math();
    const expectedMath = Math.sqrt(16) + Math.PI;
    if (Math.abs(mathResult - expectedMath) >= 0.001) {
        throw new Error(`Expected ~${expectedMath}, got ${mathResult}`);
    }
    console.log(`use_math() = ${mathResult}`);

    console.log("PASS: Native code preservation works correctly");
}

main();
